hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFs.java

```diff
@@ -1139,6 +1139,27 @@ public void mkdir(final Path dir, final FsPermission permission,
       if (theInternalDir.isRoot() && dir == null) {
         throw new FileAlreadyExistsException("/ already exits");
       }
+
+      if (this.fsState.getRootFallbackLink() != null) {
+        AbstractFileSystem linkedFallbackFs =
+            this.fsState.getRootFallbackLink().getTargetFileSystem();
+        Path p = Path.getPathWithoutSchemeAndAuthority(
+            new Path(theInternalDir.fullPath));
+        String child = (InodeTree.SlashPath.equals(dir)) ?
+            InodeTree.SlashPath.toString() :
+            dir.getName();
+        Path dirToCreate = new Path(p, child);
+        try {
+          linkedFallbackFs.mkdir(dirToCreate, permission, createParent);
+        } catch (IOException e) {
+          if (LOG.isDebugEnabled()) {
+            StringBuilder msg = new StringBuilder("Failed to create {}")
+                .append(" at fallback fs : {}");
+            LOG.debug(msg.toString(), dirToCreate, linkedFallbackFs.getUri());
+          }
+        }
+      }
+
```

Comment:
Thanx @umamaheswararao for the work here,
* Are the changes here in viewFs.java covered anywhere in test? If not will be good to cover this change as well
* Regarding the IOE, we are logging the IOE and then throwing `readOnlyMountTable("mkdir", dir);` instead of the actual exception. Is it intended? It actually suppress the actual reason, There is a difference in behavior between the implementation here and at ViewFileSystem, in case of exception there we handle it and don't throw `readOnlyMountTable("mkdir", dir);` rather a response, though mkdirs and mkdir behave differently and returning false there still might work, but here a person can not fix the issue based on the end exception he receives.

* I think the `createParent` needs to be tackled, when the parent is there in the mount table, in that case we have to explicitly change it to true. I tried tweaking one of your test give a check :
`
 @ Test
  public void testMkdirsOfDeepTreeWithFallbackLinkAndMountPathMatchingDirExist()
      throws Exception {
    Configuration conf = new Configuration();
    conf.setBoolean(Constants.CONFIG_VIEWFS_MOUNT_LINKS_AS_SYMLINKS, false);
    ConfigUtil.addLink(conf, "/user1/hive",
        new Path(targetTestRoot.toString() + "/").toUri());
    Path fallbackTarget = new Path(targetTestRoot, "fallbackDir");
    fsTarget.mkdirs(fallbackTarget);
    ConfigUtil.addLinkFallback(conf, fallbackTarget.toUri());
    AbstractFileSystem vfs = AbstractFileSystem.get(viewFsDefaultClusterUri,
        conf);
      //user1 does not exist in fallback
      Path multipleLevelToInternalDir = new Path("/user1/test");
      Path test = Path.mergePaths(fallbackTarget, multipleLevelToInternalDir);
      assertFalse(fsTarget.exists(test));
      // Creating /user1/test
    // Parent /user1 Exists.
    assertNotNull(vfs.getFileStatus(new Path("/user1")));

    // Creating /user1/test should be success, with createParent false, as
    // parent /user1 exists.
      vfs.mkdir(multipleLevelToInternalDir,
          FsPermission.getDirDefault(),false); // This throws Exception...
      assertTrue(fsTarget.exists(test));

  }
`

Reply:
1) Yes, I agree somehow we kept Viewfs.java on lowlight, probably due t it's low usage :-) I have added tests.
2) a) This is a good point. I just checked that ViewFs#mkdir is already throwing IOE. I think we can just throw IOE out. So, that users would know what happened.
    b) Another missed was, even in positive case, I was not returning before. Added return after mkdir success.
    c) We always pass createParent as true here. We don't need to actually worry parent exist or not, because parent is exist in mount Internal dir, that's why it got resolved to InternalDirViewFS#mkdir. So, passing createParent as true always in fallback case.
3) Added tests to cover the cases with ViewFs#mkdir with createParentDir true or false, with one level and multipleLevels (means to check recursive creation. Both seems to be working.)
Let's discuss if this behavior make sense to you.